infrastructure/src/main/java/org/corfudb/infrastructure/management/FailureToleranceDecisionEngine.java

```diff
@@ -0,0 +1,54 @@
+package org.corfudb.infrastructure.management;
+
+import org.corfudb.protocols.wireprotocol.NodeView;
+import org.corfudb.runtime.view.Layout;
+
+import java.util.List;
+
+/**
+ * {@link FailureToleranceDecisionEngine} provides methods to decide about failure or healing
+ * status of Corfu servers in a given {@link Layout} and for a specific view of the cluster
+ * captured in a {@link NodeView}.
+ * Decisions are dependant on the concrete {@link FailureDiscoveryPolicy} which underlies the
+ * concrete implementations of this interface.
+ *
+ * Created by Sam Behnam on 10/19/18.
+ */
+public interface FailureToleranceDecisionEngine {
+
+    /**
+     * Get the concrete {@link FailureDiscoveryPolicy} used in the current instance of
+     * {@link FailureToleranceDecisionEngine}. This policy represents the characteristics of the
+     * underlying algorithm used for making a decision about suspecting the failed or healed
+     * status of Corfu servers.
+     *
+     * @return a concrete instance of {@link FailureDiscoveryPolicy}
+     */
+    FailureDiscoveryPolicy getFailureDiscoveryPolicy();
+
+    /**
+     * Provide a list of servers in the Corfu cluster which according to the underlying
+     * {@link FailureDiscoveryPolicy} are suspected to have been failed. The decision is made
+     * based on the given view of the cluster captured in {@link NodeView} along with the
+     * expected {@link Layout}.
+     *
+     * @param nodeView view of the Corfu cluster of servers from a client node's perspective.
```

Comment:
I'm wondering if this should be `nodeView` from the client perspective or `peerView` (PR #1522) from the server perspective? My understanding is when we are dealing with link failures, it is from servers' perspective, it that correct?

Reply:
Yes this method is ultimately interested in the Cluster View captured in the map of peer views. Currently, nodeView will contain the Server metrics and therefore the map of peer views (see 1522) . So that is the path to get to peer views from node views. That being said, once the design of peer views are final, we’ll use them.